### ISSUE TITLE
Silhoutte Shade support: Cannot read property 'posKind2'

### ISSUE THREAD

**Issue by Brklnbnc** [keybuk/homebridge-powerview#1]
Hi, I've installed your plugin and am trying to set it up. After installing and adding the Platform into my config.json file, I get this error when starting HomeBridge. 

The shade with ID 12428 is a normal shade with a battery powered motor.
The other shades I have are Silhoutte shades which can tilt. I am hoping your plugin can work with this kind of shades.

Apr 23 16:28:06 domoticz homebridge[3167]: [2018-4-23 16:28:06] [PowerView] PowerView didFinishLaunching
Apr 23 16:28:06 domoticz homebridge[3167]: [2018-4-23 16:28:06] [PowerView] Updating shades
Apr 23 16:28:06 domoticz homebridge[3167]: [2018-4-23 16:28:06] Homebridge is running on port 51843.
Apr 23 16:28:06 domoticz homebridge[3167]: [2018-4-23 16:28:06] [PowerView] Update response: 48305
Apr 23 16:28:06 domoticz homebridge[3167]: [2018-4-23 16:28:06] [PowerView] now 48305/1 = 1 (425)
Apr 23 16:28:07 domoticz homebridge[3167]: [2018-4-23 16:28:07] [PowerView] Updating existing shade: 48305
Apr 23 16:28:07 domoticz homebridge[3167]: [2018-4-23 16:28:07] [PowerView] now 48305/1 = 1 (425)
Apr 23 16:28:07 domoticz homebridge[3167]: [2018-4-23 16:28:07] [PowerView] Updating existing shade: 8026
Apr 23 16:28:07 domoticz homebridge[3167]: [2018-4-23 16:28:07] [PowerView] now 8026/1 = 1 (425)
Apr 23 16:28:07 domoticz homebridge[3167]: [2018-4-23 16:28:07] [PowerView] Updating existing shade: 64623
Apr 23 16:28:07 domoticz homebridge[3167]: [2018-4-23 16:28:07] [PowerView] now 64623/1 = 1 (425)
Apr 23 16:28:07 domoticz homebridge[3167]: [2018-4-23 16:28:07] [PowerView] Updating existing shade: 45176
Apr 23 16:28:07 domoticz homebridge[3167]: [2018-4-23 16:28:07] [PowerView] now 45176/1 = 1 (546)
Apr 23 16:28:07 domoticz homebridge[3167]: [2018-4-23 16:28:07] [PowerView] Found new shade: 12428
Apr 23 16:28:07 domoticz homebridge[3167]: /usr/local/lib/node_modules/homebridge-powerview/index.js:82
Apr 23 16:28:07 domoticz homebridge[3167]: if (shadeData.positions.posKind2 == 2) {
Apr 23 16:28:07 domoticz homebridge[3167]: ^
Apr 23 16:28:07 domoticz homebridge[3167]: TypeError: Cannot read property 'posKind2' of undefined
Apr 23 16:28:07 domoticz homebridge[3167]: at PowerViewPlatform.addShade (/usr/local/lib/node_modules/homebridge-powerview/index.js:82:26)
Apr 23 16:28:07 domoticz homebridge[3167]: at PowerViewPlatform.<anonymous> (/usr/local/lib/node_modules/homebridge-powerview/index.js:160:32)
Apr 23 16:28:07 domoticz homebridge[3167]: at Request.self.callback (/usr/local/lib/node_modules/request/request.js:188:22)
Apr 23 16:28:07 domoticz homebridge[3167]: at emitTwo (events.js:126:13)
Apr 23 16:28:07 domoticz homebridge[3167]: at Request.emit (events.js:214:7)
Apr 23 16:28:07 domoticz homebridge[3167]: at Request.<anonymous> (/usr/local/lib/node_modules/request/request.js:1171:10)
Apr 23 16:28:07 domoticz homebridge[3167]: at emitOne (events.js:116:13)
Apr 23 16:28:07 domoticz homebridge[3167]: at Request.emit (events.js:211:7)
Apr 23 16:28:07 domoticz homebridge[3167]: at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/request/request.js:1091:12)
Apr 23 16:28:07 domoticz homebridge[3167]: at Object.onceWrapper (events.js:313:30)
Apr 23 16:28:07 domoticz homebridge[3167]: at emitNone (events.js:111:20)
Apr 23 16:28:07 domoticz homebridge[3167]: at IncomingMessage.emit (events.js:208:7)
Apr 23 16:28:07 domoticz homebridge[3167]: at endReadableNT (_stream_readable.js:1055:12)
Apr 23 16:28:07 domoticz homebridge[3167]: at _combinedTickCallback (internal/process/next_tick.js:138:11)
Apr 23 16:28:07 domoticz homebridge[3167]: at process._tickCallback (internal/process/next_tick.js:180:9)

**Comment by keybuk** [comment]
Hey, thanks for checking out the plugin!  I'm still developing it at the moment, so reports of problems like this are very welcome!

I don't have any Silhoutte shades myself, but I do have Duette shades, so the plugin is very much written with multi-position shades in mind.

If you'd be willing to help provide some information, I should be able to get your shades working with it too.

Firstly can you run this command and attach the output to the bug?

```
curl http://powerview-hub.local/api/shades
```

**Comment by Brklnbnc** [comment]
Thanks for developing this plugin. I'm happy to help if you need some testing. Here are the results of the api/shades query: 

{"shadeIds":[48305,8026,64623,45176,12428,47442,13154,46656,63421,24783],"shadeData":[{"id":48305,"name":"U2NodWlmcHVpIExpbmtz","roomId":47516,"groupId":2478,"order":0,"type":23,"batteryStrength":179,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},{"id":8026,"name":"UmFhbSBM","roomId":47516,"groupId":2478,"order":1,"type":23,"batteryStrength":179,"batteryStatus":3, "positions":{"position1":425,"posKind1":1}},{"id":64623,"name":"UmFhbSBS","roomId":47516,"groupId":2478,"order":2,"type":23,"batteryStrength":179,"batteryStatus":3, "positions":{"position1":425,"posKind1":1}},{"id":45176,"name":"U2NodWlmcHVpIFJlY2h0cw==","roomId":47516,"groupId":2478,"order":3,"type":23,"batteryStrength":179,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},{"id":12428,"name":"S2V1a2VuIFJvbGdvcmRpam4=","roomId":8803,"groupId":58509,"order":4,"type":42,"batteryStrength":154,"batteryStatus":3},{"id":47442,"name":"S2FudG9vciBSb2xnb3JkaWpu","roomId":21454,"groupId":20625,"order":5,"type":42,"batteryStrength":161,"batteryStatus":3},{"id":13154,"name":"QmFka2FtZXIgUm9sZ29yZGlqbg==","roomId":23771,"groupId":45165,"order":6,"type":42,"batteryStrength":158,"batteryStatus":3},{"id":46656,"name":"U2NodWlmcHVpIEw=","roomId":8043,"groupId":1697,"order":7,"type":23,"batteryStrength":180,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},{"id":63421,"name":"U2NodWlmcHVpIFI=","roomId":8043,"groupId":1697,"order":8,"type":23,"batteryStrength":180,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":485,"posKind1":1}}]}

**Comment by keybuk** [comment]
Thanks! A couple more tests, could you grab the output from this:

```
curl http://powerview-hub.local/api/shades/12428
```

And then try this (which should jog your kitchen roller blind shade), and see what the output is:

```
curl -H 'Content-Type: application/json' -X PUT --data-raw '{"shade":{"motion":"jog"}}' http://powerview-hub.local/api/shades/12428
```

That should help figure out why the positions data is empty.


For the silhouette blinds, I'm assuming those are all of the ones marked 23 in this list?

- Schuifpui Links (23) - 48305
- Raam L (23) - 8026
- Raam R (23) - 64623
- Schuifpui Rechts (23) - 45176
- Keuken Rolgordijn (42) - 12428
- Kantoor Rolgordijn (42) - 47442
- Badkamer Rolgordijn (42) - 13154
- Schuifpui L (23) - 46656
- Schuifpui R (23) - 63421
- Terrasdeur (23) - 24783 

If so, pick one you can see, and try the same command (replacing the id as appropriate):

```
curl -H 'Content-Type: application/json' -X PUT --data-raw '{"shade":{"motion":"jog"}}' http://powerview-hub.local/api/shades/46656
```

And let me know what the output of that is.

Thanks!

**Comment by Brklnbnc** [comment]
I've tested a shade in different situations, and copied the output. Silhouette shades can only be tilted when the shade is fully closed. The variable posKind changes from 1 to 3 when the shade is tilted. 

Single position: values for position1 are 0 for fully closed and 65535 for fully open with no tilting. 

Multi-position: When the shade is tilted it has to be fully closed. The posKind value is changed into 3 and values for position1 are 0 for for fully closed and no tilt, and 27515 for fully tilted.

Some output results with screenshots from the PowerView app:

Top/down positions:

Shade fully open - https://prnt.sc/j93rew
{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}}

Shade fully closed - https://prnt.sc/j93qui
{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":0,"posKind1":1}}

Shade partially opened (no tilt) - https://prnt.sc/j93pqt
{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":9914,"posKind1":1}}

Multi-positions:

Shade closed, fully tilted
Fully tilted (open) - https://prnt.sc/j93qmq
{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":27515,"posKind1":3}

Shade closed, partially tilted - https://prnt.sc/j93qc0
{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":18526,"posKind1":3}}

Shade closed, no tilt (= fully closed) - https://prnt.sc/j93q4d
{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":0,"posKind1":3}}


I hope this is useful! :)

**Comment by keybuk** [comment]
That's super-useful, thanks - can you still grab the output of the jog and stuff, because I'd like to figure out why that shade is returning an error right now.

Also:

If the shade is closed and tilted, and you send:

```
curl -H 'Content-Type: application/json' -X PUT --data-raw '{"shade":{"positions":{"position1":65535,"posKind1":1}}}' http://powerview-hub.local/api/shades/24783
```

Does it both untilt AND open?

Likewise if the shade is fully open, and you close and tilt:

```
curl -H 'Content-Type: application/json' -X PUT --data-raw '{"shade":{"positions":{"position1":18526,"posKind1":3}}}' http://powerview-hub.local/api/shades/24783
```

Does that close the shade AND tilt it?

(If the answer is yes or not to either question, it'd help to get the output from it)

**Comment by Brklnbnc** [comment]
That is correct. Shades with number 23 are Silhouette, shades with number 42 are regular rolling blinds.

Here are the results of the requested commands: 

$ curl http://192.168.1.135/api/shades/12428

{"shade":{"id":12428,"name":"S2V1a2VuIFJvbGdvcmRpam4=","roomId":8803,"groupId":58509,"order":4,"type":42,"batteryStrength":154,"batteryStatus":3},"shadeData":[{"id":47442,"name":"S2FudG9vciBSb2xnb3JkaWpu","roomId":21454,"groupId":20625,"order":5,"type":42,"batteryStrength":161,"batteryStatus":3},{"id":13154,"name":"QmFka2FtZXIgUm9sZ29yZGlqbg==","roomId":23771,"groupId":45165,"order":6,"type":42,"batteryStrength":158,"batteryStatus":3},{"id":46656,"name":"U2NodWlmcHVpIEw=","roomId":8043,"groupId":1697,"order":7,"type":23,"batteryStrength":180,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},{"id":63421,"name":"U2NodWlmcHVpIFI=","roomId":8043,"groupId":1697,"order":8,"type":23,"batteryStrength":180,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":0,"posKind1":1}}]}


$ curl -H 'Content-Type: application/json' -X PUT --data-raw '{"shade":{"motion":"jog"}}' http://192.168.1.135/api/shades/12428
curl: option --data-raw: is unknown


Trying again, with --data instead of --data-raw:

$ curl -H 'Content-Type: application/json' -X PUT --data '{"shade":{"motion":"jog"}}' http://192.168.1.135/api/shades/12428
{"shade":{"id":12428,"name":"S2V1a2VuIFJvbGdvcmRpam4=","roomId":8803,"groupId":58509,"order":4,"type":42,"batteryStrength":154,"batteryStatus":3}}

# curl -H 'Content-Type: application/json' -X PUT --data '{"shade":{"motion":"jog"}}' http://192.168.1.135/api/shades/46656

{"shade":{"id":46656,"name":"U2NodWlmcHVpIEw=","roomId":8043,"groupId":1697,"order":7,"type":23,"batteryStrength":180,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}}}

**Comment by keybuk** [comment]
Yeah, if your curl uses `--data` rather than `--data-raw` (depends on the version) just substitute that in any command - I'm on a Mac, I think you're on Linux there?

Great!

So jogging the shade updated the position.

Can you also try one more thing on a different shade:

```
curl http://192.168.1.135/api/shades/13154
```

then

```
curl http://192.168.1.135/api/shades/13154?refresh=true
```

If the first one shows no position data, hopefully that second one will grab it without jogging the shade (saving you some battery power!)

**Comment by Brklnbnc** [comment]
"If the shade is closed and tilted, and you send:"

$ curl -H 'Content-Type: application/json' -X PUT --data '{"shade":{"positions":{"position1":65535,"posKind1":1}}}' http://192.168.1.135/api/shades/24783
{"shade":{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}}}

The shade untilts and will fully open.


"Likewise if the shade is fully open, and you close and tilt. Does that close the shade AND tilt it?"
$ curl -H 'Content-Type: application/json' -X PUT --data '{"shade":{"positions":{"position1":18526,"posKind1":3}}}' http://192.168.1.135/api/shades/24783
{"shade":{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":18526,"posKind1":3}}}

It has fully closed the shade and partially tilted it.

**Comment by keybuk** [comment]
Perfect, and the hub reports that it did that.

I think I have everything I need to add the support for these shades, so I'll work on that today, and should have something for you to test by this time tomorrow!

**Comment by Brklnbnc** [comment]
I'm happy to test a new version of the plugin tomorrow! :)

And to answer your last questions as well..

"So jogging the shade updated the position."
Correct :)

"Can you also try one more thing on a different shade"

Sure:

$ curl http://192.168.1.135/api/shades/13154
{"shade":{"id":13154,"name":"QmFka2FtZXIgUm9sZ29yZGlqbg==","roomId":23771,"groupId":45165,"order":6,"type":42,"batteryStrength":158,"batteryStatus":3},"shadeData":[{"id":46656,"name":"U2NodWlmcHVpIEw=","roomId":8043,"groupId":1697,"order":7,"type":23,"batteryStrength":180,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},{"id":63421,"name":"U2NodWlmcHVpIFI=","roomId":8043,"groupId":1697,"order":8,"type":23,"batteryStrength":180,"batteryStatus":3, "positions":{"position1":65535,"posKind1":1}},{"id":24783,"name":"VGVycmFzZGV1cg==","roomId":47516,"groupId":2478,"order":9,"type":23,"batteryStrength":181,"batteryStatus":3, "positions":{"position1":18526,"posKind1":3}}]}

This result came in really fast. I guess these results were cached.


$ curl http://192.168.1.135/api/shades/13154?refresh=true
{"shade":{"id":13154,"name":"QmFka2FtZXIgUm9sZ29yZGlqbg==","roomId":23771,"groupId":45165,"order":6,"type":42,"timedOut":false,"batteryStrength":158,"batteryStatus":3, "positions":{"position1":65534,"posKind1":1}}}

The second took like 5 seconds to show up.

"If the first one shows no position data, hopefully that second one will grab it without jogging the shade (saving you some battery power!)"

Don't worry about battery power. The Silhouette shades I'm using are really large, and therefore I have connected all of them to a power adapter. I'm only using battery packs for the regular rolling blinds.

**Comment by keybuk** [comment]
Right, the PowerView hub is super-dumb… it doesn't track anything more than the shade positions you sent it via the app (or HomeKit, since the plugin pretends to be the app).

If you picked up a remote control and moved the shades, and then asked the hub where they were, it wouldn't tell you what the remote control changed.

Both jog and refresh are ways to get the hub to actually talk to the shade; refresh just asks it directly (thus taking the time), while jog helps in situations the shade itself doesn't know where it is.

What's your HomeKit plan? Just being able to use from Siri and Home.app? Or are you going to be setting up Scenes and Automations in HomeKit?

**Comment by Brklnbnc** [comment]
Thanks for the clarification. I didn't know that jog would help the shade to figure out where it is.

I have already automated pretty much everything in my house with HomeKit. For PowerView I currently use both scenes and automations. What I have done is simple. I have created scenes in the PowerView app. Then I copied the scene ID from the API, and use the plugin homebridge-http-switch to send the request via curl to the hub. This way I have created several stateless switches for all my favorite scenes (open/close/tilt). I've created a video of the scenes I use, feel free to check it out: https://www.youtube.com/watch?v=WKIfJ1D73Rs

**Comment by keybuk** [comment]
Ah cool, I skipped over the scenes and automations in PowerView and use HomeKit itself to do scenes and automations.

My goal was to, as much as possible, just have the blinds completely automated. So when it's going to be a warm sunny day, the kitchen roller blinds come down to stop the condo getting too hot in the California sun. And then in the evening, the dining room blinds come down in stages as the sun sets, and then they go back up automatically once it's behind the hill.

**Comment by Brklnbnc** [comment]
Those would be really cool automations. I'd love to implement such a thing as well. Right now I have automated opening the shades at 7am and closing the shades 30 minutes after sunset. Also I have used contact sensors to detect if a door is open and prevent the shade from closing. The other way around, when you open the door the shade will go up automatically. Very useful.

**Comment by keybuk** [comment]
Ok, give that a try…

Your silhouette shades should now have a tilt angle slider under Details

Definitely a few things to test, in particular:
 - the slider goes from -90 to 90, is that meaningful for your shades?
 - what happens at -90, and +90 - are the values right?
 - +90 is defined as "the edge facing the window is at the bottom" - is that right for your shades?
 - dragging the slider off 0 should close the shades too and HomeKit should see that they are closed
 - dragging the shades back to open should level them out, and HomeKit should see that slider go back to 0
-  you should be able to make HomeKit scenes with the slider set, it might be a bit dumb about the shade position though - try different combinations and see what happens

And obviously, file bugs if you get crashes or errors or time outs and stuff!

Thanks

Scott

**Comment by keybuk** [comment]
A bunch of debugging of another user's issues means that these should work even better for you now! Let me know how you get on.

**Comment by Brklnbnc** [comment]
I have updated the plugin, and it works like a charm! 😀 

> the slider goes from -90 to 90, is that meaningful for your shades?
> what happens at -90, and +90 - are the values right?

I see you have changed the horizontal slider from 0 to +90 degrees. This works fine. Setting the value is very responsive, as the shade moves within a split second after changing the value. 

A previous version of the plugin crashed when setting the slider to a negative value.


> +90 is defined as "the edge facing the window is at the bottom" - is that right for your shades?

Not sure what you mean by this, but I have created a photo where the shades are at +90 https://imgur.com/a/2Ogm57y


> dragging the slider off 0 should close the shades too and HomeKit should see that they are closed

HomeKit will report the shade as being "Closed" when using the slider and setting a value from 0-90. Technically this is correct, and I do not think you should change this.



> dragging the shades back to open should level them out, and HomeKit should see that slider go back to 0

I've set the horizontal slider to 50. Then I have opened the shades to 75%. The horizontal slider is set to 0.


> you should be able to make HomeKit scenes with the slider set, it might be a bit dumb about the shade position though - try different combinations and see what happens

I've create a scene and using the horizontal slider to allow filtered daylight in. It requires to set the shade as closed, then set the horizontal slider to the required value. When creating a scene with the shade opened to a set position AND setting a value with the horizontal slider, it will result in the shade open to the set position (no tilting) and ignoring the horizontal slider.



I'll do some more testing with the plugin and will let you know if I run into any issues.

A big thank you for creating this awesome plugin!

**Comment by keybuk** [comment]
For Scenes, HomeKit sends each of the characteristic changes independently. So if you were to create a scene with a shade position, and tilt angle, the plugin gets:

- set position 50%
- set tilt angle 45&deg;

The plugin is smart enough to join these back together and send just a single change to the PowerView Hub. When one or the other is zero, that one is dropped. But when both are valid values like that, it has to pick one.

I use "whichever HomeKit sends last" as the rule for that. Interesting that for your shade it was the position that won (meaning it sent the tilt first), when I tested here (without a shade) I saw the opposite. It might not even be deterministic between scene sets, and nothing we can really do about it, since we don't know the intent of the scene.

So "set shades to closed if you want to tilt" and "set tilt to 0 if you want to set the position" works as a rule.
